File: LengthConverter.cs
// LengthConverter.cs - Конвертер длин на C# (CLI + WinForms)
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Windows.Forms;

namespace LengthConverter
{
    public static class Converter
    {
        public static Dictionary<string, double> Units = new Dictionary<string, double>
        {
            {"mm", 0.001}, {"cm", 0.01}, {"dm", 0.1}, {"m", 1.0}, {"km", 1000.0},
            {"in", 0.0254}, {"ft", 0.3048}, {"yd", 0.9144}, {"mi", 1609.344}, {"nmi", 1852.0}
        };
        public static Dictionary<string, string> UnitNames = new Dictionary<string, string>
        {
            {"mm", "Миллиметр"}, {"cm", "Сантиметр"}, {"dm", "Дециметр"},
            {"m", "Метр"}, {"km", "Километр"},
            {"in", "Дюйм"}, {"ft", "Фут"}, {"yd", "Ярд"},
            {"mi", "Миля"}, {"nmi", "Морская миля"}
        };

        public static double Convert(double value, string from, string to)
        {
            if (from == to) return value;
            double meters = value * Units[from];
            return meters / Units[to];
        }

        public static List<double> ConvertBatch(List<double> values, string from, string to)
        {
            return values.Select(v => Convert(v, from, to)).ToList();
        }

        public static List<Tuple<string, string>> GenerateTable(double start, double end, double step, string from, string to, int precision)
        {
            var rows = new List<Tuple<string, string>>();
            for (double v = start; v <= end + 1e-9; v += step)
            {
                double res = Convert(v, from, to);
                rows.Add(Tuple.Create(
                    $"{v.ToString($"F{precision}")} {UnitNames[from]}",
                    $"{res.ToString($"F{precision}")} {UnitNames[to]}"
                ));
            }
            return rows;
        }
    }

    class Program
    {
        [STAThread]
        static void Main(string[] args)
        {
            if (args.Length > 0 && args[0] == "--gui")
            {
                Application.EnableVisualStyles();
                Application.Run(new LengthConverterGUI());
                return;
            }
            // CLI (упрощённый)
            var opts = new Dictionary<string, string>();
            for (int i = 0; i < args.Length; i++)
            {
                if (args[i].StartsWith("--"))
                {
                    if (i + 1 < args.Length && !args[i + 1].StartsWith("--"))
                        opts[args[i]] = args[++i];
                    else
                        opts[args[i]] = "";
                }
            }
            try
            {
                if (opts.ContainsKey("--list"))
                {
                    Console.WriteLine("Доступные единицы:");
                    foreach (var kv in Converter.UnitNames)
                        Console.WriteLine($"  {kv.Key}: {kv.Value}");
                    return;
                }
                string from = opts.GetValueOrDefault("--from", "m");
                string to = opts.GetValueOrDefault("--to", "km");
                int precision = int.Parse(opts.GetValueOrDefault("--precision", "2"));
                if (opts.ContainsKey("--range"))
                {
                    var parts = opts["--range"].Split(',');
                    if (parts.Length != 3) throw new Exception("Формат: start,end,step");
                    double start = double.Parse(parts[0]);
                    double end = double.Parse(parts[1]);
                    double step = double.Parse(parts[2]);
                    if (step <= 0) throw new Exception("Шаг должен быть положительным");
                    var table = Converter.GenerateTable(start, end, step, from, to, precision);
                    if (opts.ContainsKey("--output"))
                    {
                        File.WriteAllLines(opts["--output"], table.Select(r => $"{r.Item1} = {r.Item2}"));
                        Console.WriteLine($"Таблица сохранена в {opts["--output"]}");
                    }
                    else
                    {
                        Console.WriteLine($"Таблица {Converter.UnitNames[from]} -> {Converter.UnitNames[to]}:");
                        foreach (var r in table) Console.WriteLine($"{r.Item1} = {r.Item2}");
                    }
                    return;
                }
                if (opts.ContainsKey("--batch"))
                {
                    var values = File.ReadAllLines(opts["--batch"]).Where(l => !string.IsNullOrWhiteSpace(l))
                                    .Select(l => double.Parse(l.Trim())).ToList();
                    var results = Converter.ConvertBatch(values, from, to);
                    if (opts.ContainsKey("--output"))
                    {
                        var lines = values.Select((v, i) => $"{v.ToString($"F{precision}")} -> {results[i].ToString($"F{precision}")}");
                        File.WriteAllLines(opts["--output"], lines);
                        Console.WriteLine($"Результаты сохранены в {opts["--output"]}");
                    }
                    else
                    {
                        for (int i = 0; i < values.Count; i++)
                            Console.WriteLine($"{values[i]:F{precision}} -> {results[i]:F{precision}}");
                    }
                    return;
                }
                if (opts.ContainsKey("--value"))
                {
                    double value = double.Parse(opts["--value"]);
                    double res = Converter.Convert(value, from, to);
                    Console.WriteLine($"{value:F{precision}} {Converter.UnitNames[from]} = {res:F{precision}} {Converter.UnitNames[to]}");
                }
                else
                {
                    Console.WriteLine("Укажите --value, --batch, --range, --list или --gui");
                }
            }
            catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
        }
    }

    // ========== GUI ==========
    public class LengthConverterGUI : Form
    {
        private TextBox valueBox, resultBox;
        private ComboBox fromBox, toBox;
        private NumericUpDown precisionUpDown;

        public LengthConverterGUI()
        {
            Text = "Конвертер длин";
            Size = new System.Drawing.Size(450, 300);
            StartPosition = FormStartPosition.CenterScreen;

            var layout = new TableLayoutPanel { Dock = DockStyle.Fill, ColumnCount = 2, RowCount = 7, Padding = new Padding(10) };
            layout.RowStyles.Add(new RowStyle(SizeType.AutoSize));
            layout.RowStyles.Add(new RowStyle(SizeType.AutoSize));
            layout.RowStyles.Add(new RowStyle(SizeType.AutoSize));
            layout.RowStyles.Add(new RowStyle(SizeType.AutoSize));
            layout.RowStyles.Add(new RowStyle(SizeType.AutoSize));
            layout.RowStyles.Add(new RowStyle(SizeType.AutoSize));
            layout.RowStyles.Add(new RowStyle(SizeType.Percent, 100));

            layout.Controls.Add(new Label { Text = "Значение:", AutoSize = true }, 0, 0);
            valueBox = new TextBox();
            layout.Controls.Add(valueBox, 1, 0);

            layout.Controls.Add(new Label { Text = "Из:", AutoSize = true }, 0, 1);
            fromBox = new ComboBox { DropDownStyle = ComboBoxStyle.DropDownList, DataSource = new BindingSource(Converter.Units.Keys.ToList(), null) };
            fromBox.SelectedItem = "m";
            layout.Controls.Add(fromBox, 1, 1);

            layout.Controls.Add(new Label { Text = "В:", AutoSize = true }, 0, 2);
            toBox = new ComboBox { DropDownStyle = ComboBoxStyle.DropDownList, DataSource = new BindingSource(Converter.Units.Keys.ToList(), null) };
            toBox.SelectedItem = "km";
            layout.Controls.Add(toBox, 1, 2);

            layout.Controls.Add(new Label { Text = "Точность:", AutoSize = true }, 0, 3);
            precisionUpDown = new NumericUpDown { Minimum = 0, Maximum = 10, Value = 2 };
            layout.Controls.Add(precisionUpDown, 1, 3);

            var convertBtn = new Button { Text = "Конвертировать" };
            convertBtn.Click += (s, e) => Convert();
            layout.Controls.Add(convertBtn, 0, 4);
            layout.SetColumnSpan(convertBtn, 2);

            layout.Controls.Add(new Label { Text = "Результат:", AutoSize = true }, 0, 5);
            resultBox = new TextBox { ReadOnly = true };
            layout.Controls.Add(resultBox, 1, 5);

            var btnPanel = new FlowLayoutPanel();
            var reverseBtn = new Button { Text = "Обратный" };
            reverseBtn.Click += (s, e) => Reverse();
            btnPanel.Controls.Add(reverseBtn);
            var tableBtn = new Button { Text = "Таблица" };
            tableBtn.Click += (s, e) => ShowTable();
            btnPanel.Controls.Add(tableBtn);
            layout.Controls.Add(btnPanel, 0, 6);
            layout.SetColumnSpan(btnPanel, 2);

            Controls.Add(layout);
        }

        private void Convert()
        {
            try
            {
                double val = double.Parse(valueBox.Text);
                string from = fromBox.SelectedItem.ToString();
                string to = toBox.SelectedItem.ToString();
                int prec = (int)precisionUpDown.Value;
                double res = Converter.Convert(val, from, to);
                resultBox.Text = $"{res.ToString($"F{prec}")} {Converter.UnitNames[to]}";
            }
            catch (Exception ex) { MessageBox.Show($"Ошибка: {ex.Message}"); }
        }

        private void Reverse()
        {
            object from = fromBox.SelectedItem;
            object to = toBox.SelectedItem;
            fromBox.SelectedItem = to;
            toBox.SelectedItem = from;
            if (!string.IsNullOrEmpty(resultBox.Text)) Convert();
        }

        private void ShowTable()
        {
            try
            {
                double val = double.Parse(valueBox.Text);
                double start = val - 10, end = val + 10, step = 1.0;
                int prec = (int)precisionUpDown.Value;
                string from = fromBox.SelectedItem.ToString();
                string to = toBox.SelectedItem.ToString();
                var table = Converter.GenerateTable(start, end, step, from, to, prec);
                var sb = new System.Text.StringBuilder();
                sb.AppendLine($"{Converter.UnitNames[from]} -> {Converter.UnitNames[to]}:");
                foreach (var r in table) sb.AppendLine($"{r.Item1} = {r.Item2}");
                MessageBox.Show(sb.ToString(), "Таблица", MessageBoxButtons.OK, MessageBoxIcon.Information);
            }
            catch (Exception ex) { MessageBox.Show($"Ошибка: {ex.Message}"); }
        }
    }
}

File: LengthConverter.java
// LengthConverter.java - Конвертер длин на Java (CLI + Swing GUI)
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.List;

public class LengthConverter {
    private static final Map<String, Double> UNITS = new LinkedHashMap<>();
    private static final Map<String, String> UNIT_NAMES = new LinkedHashMap<>();
    static {
        UNITS.put("mm", 0.001); UNIT_NAMES.put("mm", "Миллиметр");
        UNITS.put("cm", 0.01);  UNIT_NAMES.put("cm", "Сантиметр");
        UNITS.put("dm", 0.1);   UNIT_NAMES.put("dm", "Дециметр");
        UNITS.put("m", 1.0);    UNIT_NAMES.put("m", "Метр");
        UNITS.put("km", 1000.0); UNIT_NAMES.put("km", "Километр");
        UNITS.put("in", 0.0254); UNIT_NAMES.put("in", "Дюйм");
        UNITS.put("ft", 0.3048); UNIT_NAMES.put("ft", "Фут");
        UNITS.put("yd", 0.9144); UNIT_NAMES.put("yd", "Ярд");
        UNITS.put("mi", 1609.344); UNIT_NAMES.put("mi", "Миля");
        UNITS.put("nmi", 1852.0); UNIT_NAMES.put("nmi", "Морская миля");
    }

    public static double convert(double value, String from, String to) {
        if (from.equals(to)) return value;
        double meters = value * UNITS.get(from);
        return meters / UNITS.get(to);
    }

    public static List<Double> convertBatch(List<Double> values, String from, String to) {
        List<Double> res = new ArrayList<>();
        for (double v : values) res.add(convert(v, from, to));
        return res;
    }

    public static List<String[]> generateTable(double start, double end, double step, String from, String to, int precision) {
        List<String[]> rows = new ArrayList<>();
        for (double v = start; v <= end + 1e-9; v += step) {
            double res = convert(v, from, to);
            rows.add(new String[]{
                String.format("%." + precision + "f %s", v, UNIT_NAMES.get(from)),
                String.format("%." + precision + "f %s", res, UNIT_NAMES.get(to))
            });
        }
        return rows;
    }

    public static void main(String[] args) {
        if (args.length > 0 && args[0].equals("--gui")) {
            SwingUtilities.invokeLater(() -> new LengthConverterGUI().setVisible(true));
            return;
        }
        // CLI (упрощённый)
        Map<String, String> opts = new HashMap<>();
        for (int i = 0; i < args.length; i++) {
            if (args[i].startsWith("--")) {
                if (i + 1 < args.length && !args[i+1].startsWith("--")) {
                    opts.put(args[i], args[++i]);
                } else {
                    opts.put(args[i], "");
                }
            }
        }
        try {
            if (opts.containsKey("--list")) {
                System.out.println("Доступные единицы:");
                for (Map.Entry<String, String> e : UNIT_NAMES.entrySet())
                    System.out.printf("  %s: %s\n", e.getKey(), e.getValue());
                return;
            }
            String from = opts.getOrDefault("--from", "m");
            String to = opts.getOrDefault("--to", "km");
            int precision = Integer.parseInt(opts.getOrDefault("--precision", "2"));
            if (opts.containsKey("--range")) {
                String[] parts = opts.get("--range").split(",");
                if (parts.length != 3) throw new Exception("Формат: start,end,step");
                double start = Double.parseDouble(parts[0]);
                double end = Double.parseDouble(parts[1]);
                double step = Double.parseDouble(parts[2]);
                if (step <= 0) throw new Exception("Шаг должен быть положительным");
                List<String[]> table = generateTable(start, end, step, from, to, precision);
                if (opts.containsKey("--output")) {
                    PrintWriter pw = new PrintWriter(opts.get("--output"));
                    for (String[] row : table) pw.println(row[0] + " = " + row[1]);
                    pw.close();
                    System.out.println("Таблица сохранена в " + opts.get("--output"));
                } else {
                    System.out.println("Таблица " + UNIT_NAMES.get(from) + " -> " + UNIT_NAMES.get(to) + ":");
                    for (String[] row : table) System.out.println(row[0] + " = " + row[1]);
                }
                return;
            }
            if (opts.containsKey("--batch")) {
                List<Double> values = new ArrayList<>();
                try (BufferedReader br = new BufferedReader(new FileReader(opts.get("--batch")))) {
                    String line;
                    while ((line = br.readLine()) != null) {
                        line = line.trim();
                        if (!line.isEmpty()) values.add(Double.parseDouble(line));
                    }
                }
                List<Double> results = convertBatch(values, from, to);
                if (opts.containsKey("--output")) {
                    PrintWriter pw = new PrintWriter(opts.get("--output"));
                    for (int i = 0; i < values.size(); i++)
                        pw.printf("%." + precision + "f -> %." + precision + "f\n", values.get(i), results.get(i));
                    pw.close();
                    System.out.println("Результаты сохранены в " + opts.get("--output"));
                } else {
                    for (int i = 0; i < values.size(); i++)
                        System.out.printf("%." + precision + "f -> %." + precision + "f\n", values.get(i), results.get(i));
                }
                return;
            }
            if (opts.containsKey("--value")) {
                double value = Double.parseDouble(opts.get("--value"));
                double res = convert(value, from, to);
                System.out.printf("%." + precision + "f %s = %." + precision + "f %s\n", value, UNIT_NAMES.get(from), res, UNIT_NAMES.get(to));
            } else {
                System.out.println("Укажите --value, --batch, --range, --list или --gui");
            }
        } catch (Exception e) {
            System.err.println("Ошибка: " + e.getMessage());
        }
    }

    // ========== GUI ==========
    static class LengthConverterGUI extends JFrame {
        private JTextField valueField, resultField;
        private JComboBox<String> fromBox, toBox;
        private JSpinner precisionSpinner;

        public LengthConverterGUI() {
            setTitle("Конвертер длин");
            setSize(450, 300);
            setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
            setLayout(new GridBagLayout());
            GridBagConstraints gbc = new GridBagConstraints();
            gbc.insets = new Insets(5, 5, 5, 5);
            gbc.fill = GridBagConstraints.HORIZONTAL;

            gbc.gridx = 0; gbc.gridy = 0;
            add(new JLabel("Значение:"), gbc);
            gbc.gridx = 1;
            valueField = new JTextField(10);
            add(valueField, gbc);

            gbc.gridx = 0; gbc.gridy = 1;
            add(new JLabel("Из:"), gbc);
            gbc.gridx = 1;
            fromBox = new JComboBox<>(UNITS.keySet().toArray(new String[0]));
            fromBox.setSelectedItem("m");
            add(fromBox, gbc);

            gbc.gridx = 0; gbc.gridy = 2;
            add(new JLabel("В:"), gbc);
            gbc.gridx = 1;
            toBox = new JComboBox<>(UNITS.keySet().toArray(new String[0]));
            toBox.setSelectedItem("km");
            add(toBox, gbc);

            gbc.gridx = 0; gbc.gridy = 3;
            add(new JLabel("Точность:"), gbc);
            gbc.gridx = 1;
            precisionSpinner = new JSpinner(new SpinnerNumberModel(2, 0, 10, 1));
            add(precisionSpinner, gbc);

            JButton convertBtn = new JButton("Конвертировать");
            convertBtn.addActionListener(e -> convert());
            gbc.gridx = 0; gbc.gridy = 4; gbc.gridwidth = 2;
            add(convertBtn, gbc);

            gbc.gridy = 5;
            add(new JLabel("Результат:"), gbc);
            gbc.gridx = 1;
            resultField = new JTextField(15);
            resultField.setEditable(false);
            add(resultField, gbc);

            JPanel btnPanel = new JPanel(new FlowLayout());
            JButton reverseBtn = new JButton("Обратный");
            reverseBtn.addActionListener(e -> reverse());
            btnPanel.add(reverseBtn);
            JButton tableBtn = new JButton("Таблица");
            tableBtn.addActionListener(e -> showTable());
            btnPanel.add(tableBtn);
            gbc.gridy = 6; gbc.gridx = 0; gbc.gridwidth = 2;
            add(btnPanel, gbc);
        }

        private void convert() {
            try {
                double val = Double.parseDouble(valueField.getText());
                String from = (String) fromBox.getSelectedItem();
                String to = (String) toBox.getSelectedItem();
                int prec = (Integer) precisionSpinner.getValue();
                double res = LengthConverter.convert(val, from, to);
                resultField.setText(String.format("%." + prec + "f %s", res, UNIT_NAMES.get(to)));
            } catch (Exception ex) {
                JOptionPane.showMessageDialog(this, "Ошибка: " + ex.getMessage());
            }
        }

        private void reverse() {
            Object from = fromBox.getSelectedItem();
            Object to = toBox.getSelectedItem();
            fromBox.setSelectedItem(to);
            toBox.setSelectedItem(from);
            if (!resultField.getText().isEmpty()) convert();
        }

        private void showTable() {
            try {
                double val = Double.parseDouble(valueField.getText());
                double start = val - 10, end = val + 10, step = 1.0;
                int prec = (Integer) precisionSpinner.getValue();
                String from = (String) fromBox.getSelectedItem();
                String to = (String) toBox.getSelectedItem();
                List<String[]> table = LengthConverter.generateTable(start, end, step, from, to, prec);
                StringBuilder sb = new StringBuilder();
                sb.append(UNIT_NAMES.get(from)).append(" -> ").append(UNIT_NAMES.get(to)).append("\n");
                for (String[] row : table) sb.append(row[0]).append(" = ").append(row[1]).append("\n");
                JTextArea area = new JTextArea(sb.toString(), 15, 40);
                area.setEditable(false);
                JOptionPane.showMessageDialog(this, new JScrollPane(area), "Таблица", JOptionPane.PLAIN_MESSAGE);
            } catch (Exception ex) {
                JOptionPane.showMessageDialog(this, "Ошибка: " + ex.getMessage());
            }
        }
    }
}
